src/rust/ide/lib/view/src/graph_editor/component/project_name.rs

```diff
@@ -0,0 +1,314 @@
+        let logger                = Logger::new("ProjectName");
+        let display_object        = display::object::Instance::new(&logger);
+        let font                  = scene.fonts.get_or_load_embedded_font("DejaVuSansMono").unwrap();
+        let size                  = Vector2::new(600.0,100.0);
+        let base_color            = DARK_GRAY_TEXT_COLOR;
+        let text_size             = 16.0;
+        let text_field_properties = TextFieldProperties{base_color,font,size,text_size};
+        let text_field            = TextField::new(&world,text_field_properties);
+        let view_logger           = Logger::sub(&logger,"view_logger");
+        let view                  = component::ShapeView::<shape::Shape>::new(&view_logger,scene);
+        let project_name          = Rc::new(RefCell::new(UNKNOWN_PROJECT_NAME.to_string()));
+        let renamed_output        = frp.outputs.renamed.clone();
+        let animations            = Animations::new(&frp.network);
+        Self{logger,view,display_object,text_field,project_name,renamed_output
+            ,animations}.initialize()
+    }
+
+    fn setup_center_alignment(&self) {
+        let mut width = 0.0;
+        self.text_field.with_mut_content(|content| {
+            let mut line = content.line(0);
+            if line.len() > 0 {
+                width = line.get_char_x_position(line.len() - 1);
+            }
+        });
+        let offset = Vector3::new(-width/2.0,0.0,0.0);
+        let mut height = default();
+            self.text_field.with_content(|content| height = content.line_height);
+        self.view.shape.sprite.size.set(Vector2::new(width,height));
+        self.view.set_position(Vector3::new(0.0,-height/2.0,0.0));
+        self.animations.positioning.set_target_value(offset);
+    }
+
+    fn initialize(self) -> Self {
+        self.add_child(&self.text_field.display_object());
+        self.add_child(&self.view);
+        let project_name = self.clone_ref();
+        self.text_field.set_text_edit_callback(move |change| {
+            // If the text edit callback is called, the TextEdit must be still alive.
+            let field_content = project_name.text_field.get_content();
+            let new_name      = field_content.replace("\n","");
+            if change.inserted == "\n" {
+                project_name.rename(&new_name);
+            }
+            // Keep only one line.
+            project_name.text_field.set_content(&new_name);
+            project_name.setup_center_alignment();
+        });
+
+        self.setup_text_field_content();
+        self
+    }
+
+    fn reset_name(&self) {
+        info!(self.logger, "Resetting project name.");
+        self.setup_text_field_content();
+    }
+
+    fn setup_text_field_content(&self) {
+        self.text_field.set_content(&self.project_name.borrow());
+        self.setup_center_alignment();
+    }
+
+    fn highlight_text(&self) {
+        self.animations.highlight.set_target_value(1.0);
+    }
```

Comment:
imo such functions should be just part of the frp definition - when reading frp block it would be easier to understand what it does instead of jumping to "model.something" one-liners. Its good tokeep in model more complex things, but things like just controlling animations, should be embedded directly in frp block so it would be clear what is happening there.

Reply:
Yeah, I thought about doing that, but keeping it like `eval_ model.view.events.mouse_over(model.fade_in_text());` is more descriptive.

Reply:
Ok, I get this rerasoning. And although I really like it, I believe that writing `alpha.set_target_value(FADE_OUT_ALPHA)` is not much worse. I really think that in case of emitting events from within eval, we should keep it as part of frp for clearness purposes.

Reply:
My problem is with its counterpart:

```rust
    fn fade_out_text(&self) {
        if !self.text_field.is_focused() {
            self.animations.opacity.set_target_value(0.0);
        }
    }
```

`fn fade_out_text` has more logic to it. The current `TextField` doesn't implement focus events and I have to poll for its state. I will keep both functions in there with a `TODO` note to make use of the newest `TextField`'s frp.